src/polyfills/custom-elements.js

```diff
@@ -0,0 +1,599 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @typedef {{
+ *   promise: !Promise<undefined>,
+ *   resolve: function(),
+ * }}
+ */
+let DeferredDef;
+
+/**
+ * @typedef {!Function}
+ */
+let CustomElementConstructorDef;
+
+/**
+ * @typedef {{
+ *  name: string,
+ *  ctor: !CustomElementConstructorDef,
+ *  observedAttributes: !Array<string>,
+ *  connectedCallback: (function()|null),
+ *  disconnectedCallback: (function()|null),
+ * }}
+ */
+let CustomElementDef;
+
+/**
+ * Validates the custom element's name.
+ * This intentionally ignores "valid" higher Unicode Code Points.
+ * https://html.spec.whatwg.org/multipage/custom-elements.html#valid-custom-element-name
+ */
+const VALID_NAME = /^[a-z][a-z0-9._]*-[a-z0-9._-]*$/;
+const INVALID_NAMES = [
+  'annotation-xml',
+  'color-profile',
+  'font-face',
+  'font-face-src',
+  'font-face-uri',
+  'font-face-format',
+  'font-face-name',
+  'missing-glyph',
+];
+
+/**
+ * Asserts that the custom element name conforms to the spec.
+ * @param {!Function} SyntaxError
+ * @param {string} name
+ */
+function assertValidName(SyntaxError, name) {
+  if (!VALID_NAME.test(name) || INVALID_NAMES.indexOf(name) >= 0) {
+    throw new SyntaxError(`invalid custom element name "${name}"`);
+  }
+}
+
+/**
+ * Does win have a full Custom Elements registry?
+ * @param {!Window} win
+ * @return {boolean}
+ */
+function hasCustomElements(win) {
+  const {customElements} = win;
+
+  return !!(
+    customElements &&
+    customElements.define &&
+    customElements.get &&
+    customElements.whenDefined);
+}
+
+/**
+ * Was HTMLElement already patched this window?
+ * @param {!Window} win
+ * @return {boolean}
+ */
+function isPatched(win) {
+  const tag = win.HTMLElement.toString();
+  return tag.indexOf('[native code]') === -1;
```

Comment:
Not full proof, but I don't think it's likely someone will obscure the toString of the `HTMLElement` constructor. I wanted to avoid doing a creation test using a subclass of the constructor, because it would require polluting the custom elements registry.

I can try thinking of a better solution if you're not comfortable with it.

Reply:
Magic property on polyfilled class?

Reply:
I was trying to allow PWA publishers to include a fully featured Custom Elements for their own code.

If they do, it's likely they'll patch `HTMLElement` as part of the polyfill. In that case, wrapping the `HTMLElement` constructor is unnecessary. But I can't do a full feature test to see if `HTMLElement` will work in our case unless I register my own CE and try `document.createElement` to see if it'll throw. This seemed like a good enough approach.